pkg/controller/registryinfra/registryinfra_controller.go

```diff
@@ -0,0 +1,696 @@
+		c.logger.Errorf("Could not sync caches for %s controller", controllerName)
+		return
+	}
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+
+	<-stopCh
+}
+
+func (c *Controller) enqueue(cluster metav1.Object) {
+	key, err := controller.KeyFunc(cluster)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %#v: %v", cluster, err))
+		return
+	}
+
+	c.queue.Add(key)
+}
+
+// worker runs a worker thread that just dequeues items, processes them, and marks them done.
+// It enforces that the syncHandler is never invoked concurrently with the same key.
+func (c *Controller) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *Controller) processNextWorkItem() bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+
+	err := c.syncHandler(key.(string))
+	c.handleErr(err, key)
+
+	return true
+}
+
+func (c *Controller) handleErr(err error, key interface{}) {
+	if err == nil {
+		c.queue.Forget(key)
+		return
+	}
+
+	logger := c.logger.WithField("cluster", key)
+
+	logger.Errorf("error syncing registry infra: %v", err)
+	if c.queue.NumRequeues(key) < maxRetries {
+		logger.Errorf("retrying registry infra")
+		c.queue.AddRateLimited(key)
+		return
+	}
+
+	utilruntime.HandleError(err)
+	logger.Infof("dropping cluster out of the queue: %v", err)
+	c.queue.Forget(key)
+}
+
+func (c *Controller) syncRegistryInfra(key string) error {
+	startTime := time.Now()
+	c.logger.WithField("key", key).Debug("syncing registry infra")
+	defer func() {
+		c.logger.WithFields(log.Fields{
+			"key":      key,
+			"duration": time.Now().Sub(startTime),
+		}).Debug("finished syncing registry infra")
+	}()
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	capicluster, err := c.clusterLister.Clusters(namespace).Get(name)
+	if err != nil {
+		return fmt.Errorf("error retrieving cluster: %v", err)
+	}
+
+	cluster, err := controller.ConvertToCombinedCluster(capicluster)
+	if err != nil {
+		return fmt.Errorf("failure converting to combined cluster: %v", err)
+	}
+
+	clusterDeploymentStatus, err := controller.ClusterProviderStatusFromCluster(capicluster)
+	if err != nil {
+		return fmt.Errorf("error retrieving clusterdeploymentstatus: %v", err)
+	}
+
+	// check if the cluster is being deleted
+	if capicluster.DeletionTimestamp != nil {
+		c.logger.Debugf("cleaning up registry infrastructure pieces for cluster %v", capicluster.Name)
+
+		_, err = c.setupRegistryInfra(cluster, false /* createInfra */)
+		if err != nil {
+			return fmt.Errorf("error deleting registry infra: %v", err)
+		}
+		err = c.deleteFinalizer(capicluster)
+		return err
+	}
+
+	if !hasRegistryInfraFinalizer(capicluster) {
+		c.logger.Debugf("adding registryinfra finalizer to cluster %v", capicluster.Name)
+		return c.addFinalizer(capicluster)
+	}
+
+	if !clusterDeploymentStatus.RegistryInfraInstalled {
```

Comment:
One thing to consider is how this was done for adding hosts to the load balencers, we don't want to do it every time we sync a machine, but we do want to periodically make sure it's as we expect.

https://github.com/openshift/cluster-operator/blob/master/pkg/controller/awselb/aws_elb_controller.go#L296

Reply:
i'll rewrite the flow so that it can always run, and the only thing that will not re-run is the credentials creation if we already have existing credentials in the secret